### PR TITLE
implement new, faster solving algorithm

### PR DESCRIPTION
**Pull request by Emerentius**
Adapted my solver to the library (after simplifying it a bit, making it faster for hard sudokus, but slower for easy ones)
- new algorithm: exact cover solver
- multiple solution finding
- added Index and IndexMut implementations for Sudoku struct (by cell
  number)
- added iter() api
- changes to the Sudoku api regarding solution finding
  - removed `.fast_solve()` and `.brute_force_solve()` because this api is leaking implementation details and the new backtracker, while not the fastest in the world, is not slow.
  - added `.solve_one()`:  
    Return first solution found
  - added `solve_unique()`:  
    Return the unique solution, if such a unique solution exists
  - `solve_at_most(limit)`:  
    Return all solutions, but at most `limit`. I have not added a function without such a limit, because I consider that a footgun. The solver would run forever if given a sufficiently underdetermined sudoku, such as an empty one.

[breaking change] due to api changes

I've bumped the version to 0.2.0 because of that and taking the liberty of adding myself to the authors.

I'm not entirely content with the solver api. I'd prefer it to return something like `Result<Sudoku, SudokuError>` instead of `Option<Sudoku>`. If the sudoku is unsolvable, one could return an Error that tells the user why the sudoku is unsolvable or why there's no unique solution returned (Unsolvable or not unique?). Something like `DuplicateNumberInRow{ row: 5, num: 3, occurences: 3 }`, `UnfillableCell { row: 3, col: 5 }`, simply `Unsolvable` when unknown.

Benchmarks:
old:

```
test easy_sudokus   ... bench: 103,528,969 ns/iter (+/- 5,282,725)
test medium_sudokus ... bench: 768,895,289 ns/iter (+/- 115,197,475)
```

new:

```
test easy_sudokus_solve    ... bench:   9,669,651 ns/iter (+/- 60,811)     (10.7x as fast)
test medium_sudokus_solve  ... bench:  13,569,888 ns/iter (+/- 156,825)    (56.7x as fast)
test hard_sudokus_solve    ... bench: 131,855,621 ns/iter (+/- 1,557,065)  (? as fast)
```

benchmarks for new apis

```
test easy_sudokus_solve_at_most_100   ... bench:  10,043,897 ns/iter (+/- 85,812)
test easy_sudokus_solve_one           ... bench:   9,195,119 ns/iter (+/- 85,663)
test easy_sudokus_solve_unique        ... bench:   9,557,791 ns/iter (+/- 68,854)
test medium_sudokus_solve_at_most_100 ... bench:  24,144,977 ns/iter (+/- 141,047)
test medium_sudokus_solve_one         ... bench:  13,570,167 ns/iter (+/- 84,067)
test medium_sudokus_solve_unique      ... bench:  13,919,837 ns/iter (+/- 167,452)
test hard_sudokus_solve_at_most_100   ... bench: 678,461,396 ns/iter (+/- 1,738,839)
test hard_sudokus_solve_one           ... bench: 131,674,409 ns/iter (+/- 1,049,071)
test hard_sudokus_solve_unique        ... bench: 235,096,577 ns/iter (+/- 3,475,457)
```

### CODE REVIEW

**Comment by Emerentius** [comment]
I can squash the commits before merge, but I'll keep them distinct for now

**Comment by aochagavia** [comment]
Awesome! Thanks for this!

File: src/sudoku.rs [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+
+		// deduce entries, but check in between deductions if the sudoku is still possible
+		while self.insert_deduced_entries() {
+			if self.is_impossible() { return }
+		}
+
+		// impossible to insert another number
+		// either solved or unsolvable
+		if self.covers.is_empty() && self.is_solved() {
+			if self.is_solved() {
```
Comment (aochagavia) [inline src/sudoku.rs:249]: Hi there! I was reading the code from your PR and noticed that the `self.is_solved()` condition is checked twice. Is this intentional? Or is it an error?

Comment (Emerentius) [reply src/sudoku.rs:249]: Looks like I wanted to remove the nesting and stopped halfway.
Yeah it is an error. `is_solved()` has no side effects.

**Pull request merged**